app/src/main/java/com/dmonzonis/nookpendium/CollectionActivity.kt

```diff
@@ -79,64 +83,87 @@ class CollectionActivity : AppCompatActivity(), SortDialogFragment.SortDialogLis
+    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
+        // Close all drawers before to avoid having multiple drawers opening on top of each other
+        closeAllDrawers()
+        return if (drawerToggle.onOptionsItemSelected(item)) {
             true
-        } else {
-            cardviewFilterThisMonth.visibility = View.GONE
-            cardviewFilterClear.visibility = View.GONE
-            cardviewSortBy.visibility = View.GONE
-            fabFilterThisMonth.hide()
-            fabFilterClear.hide()
-            fabSortBy.hide()
-            fabFilters.setImageResource(R.drawable.ic_search_black_24dp)
-            false
+        } else when (item?.itemId) {
+            R.id.miFilterButton -> {
+                if (drawerLayout.isDrawerOpen(navFilterMenu))
+                    drawerLayout.closeDrawer(navFilterMenu)
+                else
+                    drawerLayout.openDrawer(navFilterMenu)
+                true
+            }
+            else -> super.onOptionsItemSelected(item)
         }
     }
 
-    private fun filterByThisMonth() {
-        val currentMonth = Calendar.getInstance().get(Calendar.MONTH)
-        val records = recordset.applyFilterByMonth(currentMonth)
-        viewAdapter.setRecords(records)
+    private fun closeAllDrawers() {
+        drawerLayout.closeDrawer(navGamesView)
+        drawerLayout.closeDrawer(navFilterMenu)
     }
 
-    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
-        if (drawerToggle.onOptionsItemSelected(item)) {
-            return true
-        }
-        return super.onOptionsItemSelected(item)
+    private fun addFilterField(id: Int, stringId: Int) {
+        // Add filter checkbox view
+        var checkboxView = layoutInflater.inflate(R.layout.filter_checkbox, navFilterMenu, false)
+        checkboxView.id = id
+        filterContainer.addView(checkboxView)
+        var filterCheckbox = navFilterMenu.findViewById<CheckBox>(id)
+        filterCheckbox?.text = getString(stringId)
+
+        // Notify the controller of the new filter
+        if (filterCheckbox != null)
+            filterManager.addFilterCheckbox(filterCheckbox)
+    }
+
+    private fun addSortField(id: Int, stringId: Int) {
+        var sortItemLayout = layoutInflater.inflate(R.layout.sort_item, navFilterMenu, false)
+        sortItemLayout.id = id
+        sortContainer.addView(sortItemLayout)
+        var sortItem = navFilterMenu.findViewById<LinearLayout>(id)
+        var textSortBy = sortItem?.findViewById<TextView>(R.id.textSortBy)
+        textSortBy?.text = getString(stringId)
+
+        // Notify the controller of the new sort item
+        if (sortItem != null)
+            sortManager.addSortField(sortItem)
+    }
+
+    private fun setUpFilterManager() {
+        addFilterField(R.id.filterNotCaught, R.string.not_caught)
+        addFilterField(R.id.filterNotDonated, R.string.not_donated)
+        addFilterField(R.id.filterThisMonth, R.string.filter_this_month)
```

Comment:
they are being set up in the view. if you mean in the XML file, it's to avoid code repetition, instead I inflate from a premade layout and only change some parameters